libraries/hotstuff/test/hotstuff_tools.cpp

```diff
@@ -49,8 +54,189 @@ BOOST_AUTO_TEST_CASE(view_number_tests) try {
+      BOOST_CHECK_EQUAL(qc._state, state_t::unrestricted);
+
+      // add a weak vote
+      // ---------------
+      weak_vote(qc, digest, 0);
+      BOOST_CHECK_EQUAL(qc._state, state_t::weak_achieved);
+      BOOST_CHECK(qc.is_quorum_met());
+
+      // add a strong vote
+      // -----------------
+      strong_vote(qc, digest, 1);
+      BOOST_CHECK_EQUAL(qc._state, state_t::strong);
+      BOOST_CHECK(qc.is_quorum_met());
+   }
+
+   {
+      pending_quorum_certificate qc(2, 1); // 2 finalizers, quorum = 1
+      BOOST_CHECK_EQUAL(qc._state, state_t::unrestricted);
+
+      // add a strong vote
+      // -----------------
+      strong_vote(qc, digest, 1);
+      BOOST_CHECK_EQUAL(qc._state, state_t::strong);
+      BOOST_CHECK(qc.is_quorum_met());
+
+      // add a strong vote
+      // -----------------
+      strong_vote(qc, digest, 1);
+      BOOST_CHECK_EQUAL(qc._state, state_t::strong);
+      BOOST_CHECK(qc.is_quorum_met());
+   }
+
+   {
+      pending_quorum_certificate qc(3, 2); // 3 finalizers, quorum = 2
+
+      // add a weak vote
+      // ---------------
+      weak_vote(qc, digest, 0);
+      BOOST_CHECK_EQUAL(qc._state, state_t::unrestricted);
+      BOOST_CHECK(!qc.is_quorum_met());
+
+      // add a strong vote
+      // -----------------
+      strong_vote(qc, digest, 1);
+      BOOST_CHECK_EQUAL(qc._state, state_t::weak_achieved);
+      BOOST_CHECK(qc.is_quorum_met());
+
+      {
+         pending_quorum_certificate qc2(qc);
+
+         // add a weak vote
+         // ---------------
+         weak_vote(qc2, digest, 2);
+         BOOST_CHECK_EQUAL(qc2._state, state_t::weak_final);
+         BOOST_CHECK(qc2.is_quorum_met());
+      }
+
+      {
+         pending_quorum_certificate qc2(qc);
+
+         // add a strong vote
+         // -----------------
+         strong_vote(qc2, digest, 2);
+         BOOST_CHECK_EQUAL(qc2._state, state_t::strong);
+         BOOST_CHECK(qc2.is_quorum_met());
+      }
+   }
+
+   {
+      pending_quorum_certificate qc(3, 2); // 3 finalizers, quorum = 2
+
+      // add a weak vote
+      // ---------------
+      weak_vote(qc, digest, 0);
+      BOOST_CHECK_EQUAL(qc._state, state_t::unrestricted);
+      BOOST_CHECK(!qc.is_quorum_met());
+
+      // add a weak vote
+      // ---------------
+      weak_vote(qc, digest, 1);
+      BOOST_CHECK_EQUAL(qc._state, state_t::weak_final);
+      BOOST_CHECK(qc.is_quorum_met());
+
+      {
+         pending_quorum_certificate qc2(qc);
+
+         // add a weak vote
+         // ---------------
+         weak_vote(qc2, digest, 2);
+         BOOST_CHECK_EQUAL(qc2._state, state_t::weak_final);
+         BOOST_CHECK(qc2.is_quorum_met());
+      }
+
+      {
+         pending_quorum_certificate qc2(qc);
+
+         // add a strong vote
+         // -----------------
+         strong_vote(qc2, digest, 2);
+         BOOST_CHECK_EQUAL(qc2._state, state_t::weak_final);
+         BOOST_CHECK(qc2.is_quorum_met());
+      }
+   }
+
```

Comment:
Those tests are great. You can add one for invalid index.

Reply:
Hum, I have an `assert()` there, not a `EOS_ASSERT`, so I don't think I can add a test for it.  
When calling `qc.add_vote()`, we already did a lookup in our finalizer list and matched the public key, so if the index is out of range in `qc.add_vote()`, we have a serious logic problem (cannot be a user or config error), and I think that an `assert` is warranted.